timemachine/cpp/src/kernels/k_harmonic_angle_stable.cuh

```diff
@@ -0,0 +1,91 @@
+#include "../fixed_point.hpp"
+#include "k_fixed_point.cuh"
+
+template <typename RealType>
+void __global__ k_harmonic_angle_stable(
+    const int A,                        // number of angles
+    const double *__restrict__ coords,  // [N, 3]
+    const double *__restrict__ params,  // [P, 3]
+    const int *__restrict__ angle_idxs, // [A, 3]
+    unsigned long long *__restrict__ du_dx,
+    unsigned long long *__restrict__ du_dp,
+    unsigned long long *__restrict__ u) {
+
+    const auto a_idx = blockDim.x * blockIdx.x + threadIdx.x;
+
+    if (a_idx >= A) {
+        return;
+    }
+
+    int i_idx = angle_idxs[a_idx * 3 + 0];
+    int j_idx = angle_idxs[a_idx * 3 + 1];
+    int k_idx = angle_idxs[a_idx * 3 + 2];
+
+    RealType rij[3];
+    RealType rkj[3];
+
+    for (int d = 0; d < 3; d++) {
+        rij[d] = coords[j_idx * 3 + d] - coords[i_idx * 3 + d];
+        rkj[d] = coords[j_idx * 3 + d] - coords[k_idx * 3 + d];
+    }
+
+    const int ka_idx = a_idx * 3 + 0;
+    const int a0_idx = a_idx * 3 + 1;
+    const int eps_idx = a_idx * 3 + 2;
+
+    RealType ka = params[ka_idx];
+    RealType a0 = params[a0_idx];
+    RealType eps = params[eps_idx];
+
+    RealType eps2 = eps * eps;
+
+    RealType rij_dot_rkj = eps2;
+    RealType rij_dot_rij = eps2;
+    RealType rkj_dot_rkj = eps2;
+
+    for (int d = 0; d < 3; d++) {
+        rij_dot_rkj += rij[d] * rkj[d];
+        rij_dot_rij += rij[d] * rij[d];
+        rkj_dot_rkj += rkj[d] * rkj[d];
+    }
+
+    RealType norm = sqrt(rij_dot_rij * rkj_dot_rkj);
+    RealType delta = rij_dot_rkj / norm - cos(a0);
+
+    RealType cij = rij_dot_rkj / rij_dot_rij;
+    RealType ckj = rij_dot_rkj / rkj_dot_rkj;
+
+    RealType c = ka * delta / norm;
+
+    if (du_dx) {
+        for (int d = 0; d < 3; d++) {
+            RealType g_i = cij * rij[d] - rkj[d];
+            atomicAdd(du_dx + i_idx * 3 + d, FLOAT_TO_FIXED_BONDED<RealType>(c * g_i));
+
+            // Use __dadd_rd instead of (+) operator to prevent FMA
+            // optimization, which breaks bitwise equivalence for the
+            // symmetry (i, j, k) -> (k, j, i).
+            RealType g_j = __dadd_rd((1 - cij) * rij[d], (1 - ckj) * rkj[d]);
+            atomicAdd(du_dx + j_idx * 3 + d, FLOAT_TO_FIXED_BONDED<RealType>(c * g_j));
```

Comment:
1. Is there an alternative that preserves bitwise equivalence but avoids intrinsics? (e.g. converting to fixed-point first? But this may introduce worse precision loss...)

2. If unavoidable, would it be useful to extract float and double versions of this intrinsic to `gpu_utils` : `strict_add(a, b) = __dadd_rn(a, b)` or `__fadd_rn(a, b)`, `strict_mul(a, b) = __fmul_rn(a, b)` or `__dmul_rn(a, b)` (as in https://github.com/proteneer/timemachine/pull/599/files#diff-260d713349dc8e00febdb5463a13b489219f40047075484128fde22bfd65adb6 )

(Just looks odd to template on RealType = float or double, but use an intrinsic specialized to double)

3. Nit: is the rounding mode intentional? (Round-nearest mode is used elsewhere in timemachine when intrinsics are necessary to prevent formation of FMAs, but this function uses round-down)

Reply:
1. Possibly, but I haven't found it after spending some time looking. I don't think converting to fixed-point before the final multiplication is the right approach, since it seems like this would exacerbate precision loss. (I think the situation here is similar to the cross product in the periodic torsion potential, where we resort to intrinsics to ensure bitwise equality: https://github.com/proteneer/timemachine/blob/5c8a0e1cd315cd6f8e4f6481f52efca8389ffa15/timemachine/cpp/src/kernels/k_periodic_torsion.cuh#L13)
2. Good call; I think it's incorrect as-is (not sure why the tests still pass in single-precision mode). Willfix.
3. Willfix

Reply:
> e.g. converting to fixed-point first?

Actually I think this is possible without losing precision if we do something like:

```python
            unsigned long long g_j1 = FLOAT_TO_FIXED_BONDED(c * (1 - cij) * rij[d]);
            unsigned long long g_j2 = FLOAT_TO_FIXED_BONDED(c * (1 - ckj) * rkj[d]);
            atomicAdd(du_dx + j_idx * 3 + d, g_j1 + g_j2);
```

I tested this, but ended up opting for using a wrapper around the intrinsic because

1. (main reason) this version no longer matches the reference implementation with high precision (I think the reference encounters roundoff error in the final floating point addition)
2. this seems a bit less efficient (though haven't benchmarked, likely insignificant)

Reply:
Fixed (2) and (3) in https://github.com/proteneer/timemachine/pull/935/commits/1b822e48f61f742b0b09cbc97416e82f8e8e90f7

Reply:
Nice!

>ended up opting for using a wrapper around the intrinsic because ...

This looks good to me